### PR TITLE
healthcheck: support TCP health check with ProxyProtocol

### PR DESCRIPTION
**Pull request by Shikugawa**
Commit Message: healthcheck: support TCP health check with ProxyProtocol
Additional Description: This PR aims to support TCP healthcheck with ProxyProtocol. This feature is useful to ensure Proxy Protocol is enabled or not from upstream servers. It works in addition to normal TCP based healthcheck. This feature is the equivalent with `check-send-proxy` provided by HAProxy. https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#5.2-check-send-proxy
Risk Level: Low
Testing: Unit/Integration
Docs Changes: Required
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lizan** [comment]
/lgtm api

@zuercher for owner review

**Comment by Shikugawa** [comment]
@zuercher friendly ping

File: api/envoy/config/core/v3/health_check.proto
Comment (zuercher) [inline api/envoy/config/core/v3/health_check.proto:185]: For the comment, would prefer: "Use the proxy protocol for health checks. The send payload is sent immediately after the proxy protocol header."

Furthermore, is it valid to have an empty `send` payload when this flag is enabled? If so we should both document that and guard against configurations that leave `send` empty.

File: source/extensions/health_checkers/tcp/health_checker_impl.cc
Comment (zuercher) [inline source/extensions/health_checkers/tcp/health_checker_impl.cc:153]: There another proxy protocol command "local" which looks like it's meant for health checks (and that the remote server should use the connection endpoints and causes the endpoints in the protocol to be ignored). I *think* it's okay to use the command this function generates since we'll set the endpoints to be ourselves the server being health checked, but I thought I'd point it out.

Comment (Shikugawa) [reply source/extensions/health_checkers/tcp/health_checker_impl.cc:153]: Yes. Actually we can use local command for filling the requirements to check whether ProxyProtocol is enabled by healthcheck. But I think that we can configure more fine-grained healthcheck strategy by setting TLVs defined from ProxyProtocol config. So I think It's better to use `generateProxyProtocolHeader` for advanced requirements.

File: api/envoy/config/core/v3/health_check.proto
Comment (Shikugawa) [reply api/envoy/config/core/v3/health_check.proto]: Sorry I forcefully pushed. The fixed commits are shown here. https://github.com/envoyproxy/envoy/pull/31691/commits/495ba97c84b7b442fc51d791bdff37ce932cd9a5#diff-4ccbbf045baa1ad3abd43bb7534d765e115e50be2ac9d9c3639588a4fd54e8a6R183-R185

**Review by zuercher — APPROVED** [review]
@ilzan sorry, I asked for a comment change in the API and it needs to be re-approved.

Otherwise, lgtm.

**Comment by lizan** [comment]
@Shikugawa one last thing, can you add a release note for the new field?

**Review by zuercher — APPROVED** [review]

File: changelogs/current.yaml
Comment (zuercher) [inline changelogs/current.yaml:30]: This line needs to end with a period to make the linter happy.

Comment (Shikugawa) [reply changelogs/current.yaml]: Thank you for pointing. Fixed https://github.com/envoyproxy/envoy/pull/31691/commits/d25971baac735e4329a8b4bce1e1c0178e939020

**Review by zuercher — DISMISSED** [review]
lgtm

File: source/extensions/health_checkers/tcp/health_checker_impl.cc
Comment (botengyao) [reply source/extensions/health_checkers/tcp/health_checker_impl.cc:153]: Could you elaborate more for the meaning of address setting in the Proxy Protocol header? The health check's local address is the upstream local address, and the remote is the health check address of the remote host. I think this is different from the real upstream request which is using the downstream connection as the input of this `generateProxyProtoHeader` util.

I'd incline to add some comments for this, and describe the `LOCAL` as well as the address setting. Also could you add a test to see if the proxy protocol with address is actually sent/received?

**Review by botengyao — COMMENTED** [review]
Thanks for working on this! Left one comment.

/wait

**Comment by zuercher** [comment]
ping @Shikugawa

File: source/extensions/health_checkers/tcp/health_checker_impl.cc
Comment (Shikugawa) [reply source/extensions/health_checkers/tcp/health_checker_impl.cc]: @botengyao @zuercher Hi. Sorry for late response.
I revisited how to provide this feature and came up with that we should provide this feature with local command for V2, and present local/remote address for V1 because it doesn't support them. This is why I found that some users faced confusion for this behavior to provide with PROXY command for healthcheck in HAProxy community. 
 https://github.com/haproxy/haproxy/issues/511

So, I cleaned entire codes in this patch to support them.

**Comment by phlax** [comment]
@botengyao @zuercher i think this is waiting for final review

**Review by zuercher — APPROVED** [review]
@lizan this needs API re-approval because a comment was updated

**Review by lizan — COMMENTED** [review]
/lgtm api

**Review by botengyao — APPROVED** [review]

**Review by botengyao — COMMENTED** [review]
/assign-from @envoyproxy/senior-maintainers 

For a pass and merging.

**Review by zuercher — DISMISSED** [review]

**Comment by Shikugawa** [comment]
Fixed conflict. Could you please approve again?

**Review by ravenblackx — APPROVED** [review]
Approving the conflict-merge commit only, and assuming zuercher's review of everything before that still applies.

**Pull request merged**
